tools/fosdd.1
.\" 
.TH "FOSDD" "1" "Mars 2023" "fosdd" "User Commands"
.SH "NAME"
fosdd \- FOS dd
.SH "SYNOPSIS"
.B fosdd
[\fIoptions\fR] \fIdevice \fIdestination
.SH "DESCRIPTION"
Tool to convert floppy image to disk image and vice-versa.
.TP 
\fB\-h\fR \fB\-\-help\fR
Display help message.
.TP 
\fB\-v\fR \fB\-\-version\fR
Show version.
.TP 
\fB\-l\fR \fB\-\-force\fR
Overwrite the destination if exists.
.TP 
\fB\-l\fR \fB\-\-fos\-logger\fR
Turn ON the FOS logger.
.TP 
\fBdevice\fR
/dev/fd0 for floppy disk
.br 
/dev/sda for hard disk, etc, ...
.br 
image.di for a floppy or disk image
.TP 
\fBdestination\fR
The destination to save the image.
.SH "AUTHOR"
Written by Mathieu Schroeter <mathieu@schroetersa.ch>.
.SH "REPORTING BUGS"
Report bugs to <\fImathieu@schroetersa.ch\fP>.
.SH "COPYRIGHT"
Copyright \(co 2023 Mathieu Schroeter

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


ChangeLog
2023-03-26  Mathieu Schroeter <mathieu@schroetersa.ch>

	* Release 1.0.0

	* fosdd: new tool that can be used in order to convert floppy device
	  or image into hard disk image and vice-versa.

	* fosmount: the PBM and XPM images use a shorter filename.

2010-03-14  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Release 0.4.0

	* Many improvements in the build scripts and Makefiles. The build for
	  Windows uses the --cross-compile option with the configure script.
	  The external scripts to build for Windows, provided with the older
	  versions are no longer necessary. Now the DLLs are created in the
	  right way.
	  The pkg-config files are now available for libfosfat and libfosgra.

	* libfosgra: add support for the decoding of .COLOR files.

	* fosmount: add the conversion on the fly of .COLOR files to .XPM files.
	  Note that the XPM files use the version 2 of X PixMap. It needs a real
	  application like The Gimp in order to read the files. Or the tool SXPM
	  from the X11 (X.Org) project.

	* smascii: add a manpage. Thanks to Didier Raboud from Debian.

	* libw32disk: this library doesn't rely anymore to C++. It is now fully
	  rewritten in C with a proper license in order to be redistribuable by
	  Debian.

2009-07-07  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New beta released 0.3.2

	* libfosgra: new library to support .IMAGE decoding.

	* fosmount: some improvements to convert on the fly the images [.IMAGE]
	  to PBM (portable bitmap format) with libfosgra.

	* Fix doxygen, packaging and Makefiles.

2009-07-04  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New beta released 0.3.1

2009-05-06  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* fosrec: Add a new tool to easily recovery files.

	* Fix Debian/Ubuntu packaging.

2009-01-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* A lot of cleaning and all doxygen documentations are only in the
	  public header file.

	* Add option in configure to generate doxygen HTML and PDF outputs.

2008-02-04  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Improve a lot the code lisibility.

	* Fix some memleaks when errors appear.

2008-01-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Add manpages for fosmount and fosread.

2007-12-16  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New major beta released 0.3.0

	* Add Window$ support, and use MinGW32 to compile DLL libraries,
	  fosread and smascii. A script package-win32.sh can be used to create
	  a zip package.

	* libfosfat: add support for Window$ through libw32disk.

	* libw32disk: add C wrapper to use original Disk.cxx class with
	  libfosfat.

	* fosread: fix symlink support, now that is possible to read like a
	  directory. Add support to list and restore deleted files.

2007-12-13  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: first implementation of an undelete tool just disabled by
	  default. The next step is added some options to list and restore
	  files.

2007-12-12  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: add support for multiple devices. A new main structure is
	  used instead of global variables. Only the logger is always global.

2007-12-09  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Fourth beta released 0.2.3

	* Some fixes, cleaned code, ...

2007-07-13  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Change license GPLv2 to GPLv3.

	* Fix a lot of memory leaks.

2007-07-07  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Third beta released 0.2.2

2007-06-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: change the auto detection function for a better way. Use
	  the name was very bad because a lot of disks have no name. Then the
	  test is does with the CHK (look the code for more informations).

2007-06-24  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Add auto detection for the disk type. The type can be forced with
	  fosmount and fosread if necessary.

	* libfosfat: an internal logger can be enabled to know what the FOS is
	  doing. A new option is available in fosread and fosmount.

2007-06-22  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Second beta released 0.2.1

	* Add support for symbolic links. The Smaky's soft links are now
	  converted to symlink.

	* libfosfat: the cache is now released when the device is closed.

2007-06-20  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: a new system of caching is implemented to improve the
	  speed. Now that will be very fast for looking in a directory. The
	  system must be fixed for that the memory is released.

	* fosmount: a bug is fixed for that FUSE is always loaded as
	  single-thread. If not there is a really a big problem when data are
	  loaded with libfosfat.

2007-05-10  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* First beta released 0.2.0

2007-04-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Use getopt header with fosmount and fosread and add --help and
	  --version options.

2007-03-25  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* smascii: add an option to convert Carriage Return to Unix newline
	  (Line Feed).

	* Fix Ubuntu package builder with CDBS.

2007-03-24  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Some fixes in all the project.

	* smascii: new tool to convert Smaky's text files to Extended ASCII
	  (ISO-8859-1) files.

2007-03-22  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: change all long variables for int because the FOS uses only
	  32 bits addresses. The lpow() function is removed and it was used only
	  for the c2l() function. Then a shift register is enough and faster.

2007-03-19  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: add a test for CHK value in BD and BL. If this value
	  changes, then NULL is returned and not the block.

tools/fosdd.c
/*
 * FOS fosdd: bidirectional tool to convert floppy (.DI) img to harddisk img
 * Copyright (C) 2023 Mathieu Schroeter <mathieu@schroetersa.ch>
 *
 * Thanks to Pierre Arnaud for his help and the documentation
 *    And to Epsitec SA for the Smaky computers
 *
 * This file is part of Fosfat.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <unistd.h>

#include "fosfat.h"
#include "fosfat_internal.h"

#define HELP_TEXT \
"Tool to convert floppy image to disk image and vice-versa. fosfat-" LIBFOSFAT_VERSION_STR "\n\n" \
"Usage: fosdd [options] device destination\n\n" \
" -h --help             this help\n" \
" -v --version          version\n" \
" -f --force            overwrite the destination if exists\n" \
" -l --fos-logger       that will turn on the FOS logger\n" \
"\n" \
"\nPlease, report bugs to <mathieu@schroetersa.ch>.\n"

#define VERSION_TEXT "fosdd-" LIBFOSFAT_VERSION_STR "\n"

/* Print help. */
static void
print_info (void)
{
  printf (HELP_TEXT);
}

/* Print version. */
static void
print_version (void)
{
  printf (VERSION_TEXT);
}

/* Add blocks at the beginning according to the disk type. */
static int
dd (fosfat_t *fosfat, fosfat_disk_t type, const char *output_file)
{
  FILE *f_out = NULL;
  f_out = fopen (output_file, "wb");
  if (!f_out)
    return -1;

  size_t offset = 0x10;

  /* The source is a floppy and the destination a disk */
  if (type == FOSFAT_FD)
  {
    /* Add padding of 0x10 (offset) */
    static const char zero[FOSFAT_BLK] = {0};
    /* Static boot entry for hard disk */
    static const char boot[FOSFAT_BLK] = {
      0x28, 0x3c, 0x80, 0x00, 0x00, 0x10, 0x7a, 0x10,
      0x70, 0x10, 0x28, 0x7c, 0x00, 0x02, 0x00, 0x00,
      0x4e, 0xb9, 0x00, 0x01, 0xe0, 0x08, 0x4a, 0x47,
      0x66, 0x00, 0x00, 0x08, 0x4e, 0xf9, 0x00, 0x02,
      0x00, 0x00, 0xa0, 0x42, 0x0d, 0x2a, 0x2a, 0x2a,
      0x2a, 0x2a, 0x20, 0x45, 0x72, 0x72, 0x65, 0x75,
      0x72, 0x20, 0x64, 0x65, 0x20, 0x6c, 0x65, 0x63,
      0x74, 0x75, 0x72, 0x65, 0x00, 0x00, 0x60, 0x00,
      0x00, 0x24, 0xa0, 0x42, 0x0d, 0x2a, 0x2a, 0x2a,
      0x2a, 0x2a, 0x20, 0x50, 0x72, 0x6f, 0x67, 0x72,
      0x61, 0x6d, 0x6d, 0x65, 0x20, 0x53, 0x54, 0x41,
      0x52, 0x54, 0x55, 0x50, 0x20, 0x61, 0x62, 0x73,
      0x65, 0x6e, 0x74, 0x00, 0xa0, 0x42, 0x2c, 0x20,
      0x64, 0x12, 0x6d, 0x61, 0x72, 0x72, 0x61, 0x67,
      0x65, 0x20, 0x69, 0x6d, 0x70, 0x6f, 0x73, 0x73,
      0x69, 0x62, 0x6c, 0x65, 0x0d, 0x0d, 0x00, 0x00,
      0x4e, 0x41, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
    };

    size_t read = 0;
    while (read < (offset * FOSFAT_BLK))
    {
      size_t to_write = (offset * FOSFAT_BLK) - read;
      if (to_write > FOSFAT_BLK)
        to_write = FOSFAT_BLK;
      fwrite (read == 0x400 ? boot : zero, 1, to_write, f_out);
      read += to_write;
    }
  }

  for (int blk = -offset;; ++blk)
  {
    fosfat_data_t *read_buffer = fosfat_read_d (fosfat, blk);
    if (!read_buffer)
      break;

    fwrite (read_buffer->data, 1, FOSFAT_BLK, f_out);
    free (read_buffer);
  }

  fclose (f_out);
  return 0;
}

int
main (int argc, char **argv)
{
  fosfat_t *fosfat;
  int res = 0, force = 0, next_option;
  fosfat_disk_t type = FOSFAT_AD;
  const char *input_file;
  const char *output_file;

  const char *const short_options = "hflv";

  const struct option long_options[] = {
    { "help",       no_argument, NULL, 'h' },
    { "force",      no_argument, NULL, 'f' },
    { "fos-logger", no_argument, NULL, 'l' },
    { "version",    no_argument, NULL, 'v' },
    { NULL,         0,           NULL,  0  }
  };

  /* check options */
  do
  {
    next_option = getopt_long (argc, argv, short_options, long_options, NULL);
    switch (next_option)
    {
    default :           /* unknown */
    case '?':           /* invalid option */
    case 'h':           /* -h or --help */
      print_info ();
      return -1;
    case 'v':           /* -v or --version */
      print_version ();
      return -1;
    case 'f':
      force = 1;
      break;
    case 'l':           /* -l or --fos-logger */
      fosfat_logger (1);
      break;
    case -1:            /* end */
      break ;
    }
  } while (next_option != -1);

  if (argc < optind + 2)
  {
    print_info ();
    return -1;
  }

  input_file  = argv[optind + 0];
  output_file = argv[optind + 1];

  if (!force && !access (output_file, F_OK)) {
    fprintf (stderr,
             "The output file \"%s\" already exists, use --force to overwrite\n",
             output_file);
    return -1;
  }

  fosfat = fosfat_open (input_file, FOSFAT_AD, 0);
  if (!fosfat)
  {
    fprintf (stderr, "Could not open %s!\n", input_file);
    return -1;
  }

  type = fosfat_type (fosfat);
  switch (type)
  {
  case FOSFAT_FD:
    printf ("Convert floppy \"%s\" into disk image at \"%s\"\n",
            input_file, output_file);
    res = dd (fosfat, type, output_file);
    break;

  case FOSFAT_HD:
    printf ("Convert disk \"%s\" into floppy image at \"%s\"\n",
            input_file, output_file);
    res = dd (fosfat, type, output_file);
    break;

  default:
    fprintf (stderr, "Input device / image not supported!\n");
    return -2;
  }

  fosfat_close (fosfat);

  if (res)
    fprintf (stderr, "Unexpected error when converting the device / image!\n");

  return res;
}
